### PR TITLE
fix(VBtn): show background on non-elevated buttons, fix color states

### LINKED ISSUE #12903: [Bug Report] Button background colors broken in 2.4.1
### Environment
**Vuetify Version:** 2.4.1
**Last working version:** 2.4.0
**Vue Version:** 2.6.11
**Browsers:** Chrome 87.0.4280.88
**OS:** Mac OS 10.15.7

### Steps to reproduce
Just create a regular button with color="primary" or any other color applied.

### Expected Behavior
Background should take the color -prop value.

### Actual Behavior
Color is applied to the button text instead.

### Reproduction Link
<a href="https://codepen.io/juhasev/pen/yLajdGo" target="_blank">https://codepen.io/juhasev/pen/yLajdGo</a>

### Other comments
I am using dark theme but it appears this bug also exists in the light theme, see the codepen supplied.

### PR DESCRIPTION
**Pull request by KaelWD**
fixes #12903

related: #12892, #12845, #12814

## Markup:

<details>

v2.3: https://codepen.io/kaelwd/pen/oNzybwv  
v2.4: https://codepen.io/kaelwd/pen/abmKdWM

```vue
<template>
  <v-container>
    baseline: <br>
    <v-btn>normal</v-btn>
    <v-btn text>text</v-btn>
    <v-btn plain>plain</v-btn>
    <v-btn depressed>depressed</v-btn>
    <v-btn outlined>outlined</v-btn>

    <hr class="my-2">

    coloured: <br>
    <v-btn color="primary">normal</v-btn>
    <v-btn
      color="primary"
      text
    >
      text
    </v-btn>
    <v-btn
      color="primary"
      plain
    >
      plain
    </v-btn>
    <v-btn
      color="primary"
      depressed
    >
      depressed
    </v-btn>
    <v-btn
      color="primary"
      outlined
    >
      outlined
    </v-btn>

    <hr class="my-2">

    elevation: <br>
    <v-btn elevation="5">
      normal
    </v-btn>
    <v-btn
      elevation="5"
      text
    >
      text
    </v-btn>
    <v-btn
      elevation="5"
      plain
    >
      plain
    </v-btn>
    <v-btn
      depressed
      elevation="5"
    >
      depressed
    </v-btn>
    <v-btn
      elevation="5"
      outlined
    >
      outlined
    </v-btn>

    <hr class="my-2">

    elevation + coloured: <br>
    <v-btn color="primary">
      normal
    </v-btn>
    <v-btn
      color="primary"
      elevation="5"
      text
    >
      text
    </v-btn>
    <v-btn
      color="primary"
      elevation="5"
      plain
    >
      plain
    </v-btn>
    <v-btn
      color="primary"
      depressed
      elevation="5"
    >
      depressed
    </v-btn>
    <v-btn
      color="primary"
      elevation="5"
      outlined
    >
      outlined
    </v-btn>

    <hr class="my-2">

    disabled: <br>
    <v-btn disabled>normal</v-btn>
    <v-btn
      disabled
      text
    >
      text
    </v-btn>
    <v-btn
      disabled
      plain
    >
      plain
    </v-btn>
    <v-btn
      depressed
      disabled
    >
      depressed
    </v-btn>
    <v-btn
      disabled
      outlined
    >
      outlined
    </v-btn>

    <hr class="my-2">

    disabled + coloured: <br>
    <v-btn
      color="primary"
      disabled
    >
      normal
    </v-btn>
    <v-btn
      color="primary"
      disabled
      text
    >
      text
    </v-btn>
    <v-btn
      color="primary"
      disabled
      plain
    >
      plain
    </v-btn>
    <v-btn
      color="primary"
      depressed
      disabled
    >
      depressed
    </v-btn>
    <v-btn
      color="primary"
      disabled
      outlined
    >
      outlined
    </v-btn>

    <hr class="my-2">

    with icon: <br>
    <v-btn>
      normal
      <v-icon
        dark
        right
      >mdi-cloud-upload
      </v-icon>
    </v-btn>
    <v-btn text>
      text
      <v-icon
        dark
        right
      >mdi-cloud-upload
      </v-icon>
    </v-btn>
    <v-btn plain>
      plain
      <v-icon
        dark
        right
      >mdi-cloud-upload
      </v-icon>
    </v-btn>
    <v-btn depressed>
      depressed
      <v-icon
        dark
        right
      >mdi-cloud-upload
      </v-icon>
    </v-btn>
    <v-btn outlined>
      outlined
      <v-icon
        dark
        right
      >mdi-cloud-upload
      </v-icon>
    </v-btn>

    <hr class="my-2">

    coloured with icon: <br>
    <v-btn color="primary">
      normal
      <v-icon
        dark
        right
      >mdi-cloud-upload
      </v-icon>
    </v-btn>
    <v-btn
      color="primary"
      text
    >
      text
      <v-icon
        dark
        right
      >mdi-cloud-upload
      </v-icon>
    </v-btn>
    <v-btn
      color="primary"
      plain
    >
      plain
      <v-icon
        dark
        right
      >mdi-cloud-upload
      </v-icon>
    </v-btn>
    <v-btn
      color="primary"
      depressed
    >
      depressed
      <v-icon
        dark
        right
      >mdi-cloud-upload
      </v-icon>
    </v-btn>
    <v-btn
      color="primary"
      outlined
    >
      outlined
      <v-icon
        dark
        right
      >mdi-cloud-upload
      </v-icon>
    </v-btn>

    <hr class="my-2">

    coloured + disabled with icon: <br>
    <v-btn
      color="primary"
      disabled
    >
      normal
      <v-icon
        dark
        right
      >mdi-cloud-upload
      </v-icon>
    </v-btn>
    <v-btn
      color="primary"
      disabled
      text
    >
      text
      <v-icon
        dark
        right
      >mdi-cloud-upload
      </v-icon>
    </v-btn>
    <v-btn
      color="primary"
      disabled
      plain
    >
      plain
      <v-icon
        dark
        right
      >mdi-cloud-upload
      </v-icon>
    </v-btn>
    <v-btn
      color="primary"
      depressed
      disabled
    >
      depressed
      <v-icon
        dark
        right
      >mdi-cloud-upload
      </v-icon>
    </v-btn>
    <v-btn
      color="primary"
      disabled
      outlined
    >
      outlined
      <v-icon
        dark
        right
      >mdi-cloud-upload
      </v-icon>
    </v-btn>

    <hr class="my-2">

    icon: <br>
    <v-btn icon>
      <v-icon>mdi-cloud-upload</v-icon>
    </v-btn>
    <v-btn
      icon
      text
    >
      <v-icon>mdi-cloud-upload</v-icon>
    </v-btn>
    <v-btn
      icon
      plain
    >
      <v-icon>mdi-cloud-upload</v-icon>
    </v-btn>
    <v-btn
      icon
      depressed
    >
      <v-icon>mdi-cloud-upload</v-icon>
    </v-btn>
    <v-btn
      icon
      outlined
    >
      <v-icon>mdi-cloud-upload</v-icon>
    </v-btn>

    <hr class="my-2">

    coloured + icon: <br>
    <v-btn
      color="primary"
      icon
    >
      <v-icon>mdi-cloud-upload</v-icon>
    </v-btn>
    <v-btn
      color="primary"
      icon
      text
    >
      <v-icon>mdi-cloud-upload</v-icon>
    </v-btn>
    <v-btn
      color="primary"
      icon
      plain
    >
      <v-icon>mdi-cloud-upload</v-icon>
    </v-btn>
    <v-btn
      color="primary"
      icon
      depressed
    >
      <v-icon>mdi-cloud-upload</v-icon>
    </v-btn>
    <v-btn
      color="primary"
      icon
      outlined
    >
      <v-icon>mdi-cloud-upload</v-icon>
    </v-btn>
  </v-container>
</template>
```
</details>

### CODE REVIEW

File: packages/vuetify/src/components/VBtn/VBtn.ts [hunk 1]
```diff
@@ -191,7 +188,7 @@ export default baseMixins.extend<options>().extend({
       this.loading && this.genLoader(),
     ]
     const { tag, data } = this.generateRouteLink()
-    const setColor = (this.isElevated || this.depressed)
```
Comment (KaelWD) [inline packages/vuetify/src/components/VBtn/VBtn.ts:194]: This condition was an exact copy of `hasBg`

File: packages/vuetify/src/components/VBtn/VBtn.ts [hunk 2]
```diff
@@ -130,7 +127,7 @@ export default baseMixins.extend<options>().extend({
         !this.depressed &&
         !this.disabled &&
         !this.plain &&
-        (this.elevation != null || Number(this.elevation) < 1)
```
Comment (KaelWD) [inline packages/vuetify/src/components/VBtn/VBtn.ts:133]: This removed the elevation from plain buttons

File: packages/vuetify/src/components/VBtn/VBtn.ts [hunk 3]
```diff
@@ -120,7 +117,7 @@ export default baseMixins.extend<options>().extend({
       else return this.ripple ?? defaultRipple
     },
     hasBg (): boolean {
-      return this.isElevated || this.depressed
+      return !this.text && !this.plain && !this.outlined && !this.icon
```
Comment (KaelWD) [inline packages/vuetify/src/components/VBtn/VBtn.ts:120]: duh

File: packages/vuetify/src/components/VBtn/VBtn.ts [hunk 4]
```diff
@@ -83,13 +83,10 @@ export default baseMixins.extend<options>().extend({
         'v-btn--absolute': this.absolute,
         'v-btn--block': this.block,
         'v-btn--bottom': this.bottom,
-        'v-btn--contained': this.isElevated, // TODO: remove v3
-        'v-btn--depressed': (this.depressed) || this.outlined, // TODO: remove v3
         'v-btn--disabled': this.disabled,
         'v-btn--is-elevated': this.isElevated,
         'v-btn--fab': this.fab,
         'v-btn--fixed': this.fixed,
-        'v-btn--flat': !this.isElevated, // TODO: remove v3,
```
Comment (KaelWD) [inline packages/vuetify/src/components/VBtn/VBtn.ts:92]: Not sure what the TODO is about, none of these classes are used any more.

File: packages/vuetify/src/components/VBtn/VBtn.sass [hunk 5]
```diff
@@ -234,11 +232,6 @@
 .v-btn--outlined
   border: $btn-outline-border-width solid currentColor
 
-.v-btn:not(.v-btn--has-bg)
-  .v-btn__content
-    .v-icon
-      color: currentColor
```
Comment (KaelWD) [inline packages/vuetify/src/components/VBtn/VBtn.sass:240]: Redundant

File: packages/vuetify/src/components/VBtn/VBtn.sass [hunk 6]
```diff
@@ -22,9 +22,6 @@
     .v-btn__loading
       color: map-deep-get($material, 'buttons', 'disabled') !important
 
-    &:not(.v-btn--outlined):not(.v-btn--plain):not(.v-btn--text)
```
Comment (KaelWD) [inline packages/vuetify/src/components/VBtn/VBtn.sass:25]: This is the same logic as `hasBg` but in css

**Review by ElijahKotyluk — APPROVED** [review]
Thank you @KaelWD

**Review by johnleider — APPROVED** [review]

**Comment by anestorov** [comment]
still not working using NPM. Using CDN its fixed

**Comment by MajesticPotatoe** [comment]
> still not working using NPM version 2.4.2. Using CDN is ok but NPM version seems not fixed

try deleting your `node_modules` folder. `npm` is notorious for caching things.

**Comment by anestorov** [comment]
> > still not working using NPM version 2.4.2. Using CDN is ok but NPM version seems not fixed
> 
> try deleting your `node_modules` folder. `npm` is notorious for caching things.

I tried everything and switched back to v 2.3.23 which works. Additionally it seems 2.4+ has different margins and paddings for containers and cards so it messes up design.

**Comment by MajesticPotatoe** [comment]
> I tried everything and switched back to v 2.3.23 which works.

[submit an issue](https://issues.vuetify.com) with a valid minimal reproduction.

> Additionally it seems 2.4+ has different margins and paddings for containers and cards so it messes up design.

this is a known issue related to nested v-row and negative margins and is not related to this issue.

**Pull request merged**